### ISSUE TITLE
508-defect-3 [TOUCH]: Touch target size should be at least 44x44px

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#23976]
# [508-defect-3](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

As a Veteran with hand tremors or mobility impairments, I want to a larger touch target to click on.

## Details

Currently, the button's size is 26x26, which is considerably smaller from WCAG's minimum recommendation of 44x44. A larger button size or a visible label may help Veterans with hand tremors or mobility impairments more accurately click or tap on it. 

## Acceptance Criteria

- [ ] Button size is at least 44x44px OR
- [ ] Text label is added

## WCAG or Vendor Guidance (optional)

[Target Size](https://dequeuniversity.com/resources/wcag2.1/2.5.5-target-size)

### Current
<img width="562" alt="Screen Shot 2021-04-28 at 4 41 01 PM" src="https://user-images.githubusercontent.com/14154792/116470442-6ec9ed80-a841-11eb-8ef9-43e617d74733.png">

## Larger touch target
<img width="566" alt="Screen Shot 2021-04-28 at 4 40 01 PM" src="https://user-images.githubusercontent.com/14154792/116470549-91f49d00-a841-11eb-82d7-651b5953e132.png">

## Visible label
Having a visible label would also help Veterans with cognitive or learning disabilities better understand what this button would do. It also gives a visible label for Veterans who rely on voice commands to select. e.g. "Click copy link button" vs. "Click... ?"

<img width="780" alt="Screen Shot 2021-04-28 at 4 47 05 PM" src="https://user-images.githubusercontent.com/14154792/116470440-6ec9ed80-a841-11eb-9557-6230c4b6963e.png">

**Comment by SKasimow** [comment]
@youngfreezyVA Hi Fareez - assigning this bug over to you. It's not a launch-blocker so nice-to-have done for tomorrow's staging accessibility review but not the priority. Thanks, Sharon

cc @jenniferlee-dsva

**Comment by jenniferlee-dsva** [comment]
Is there a way to control the size of the tappable area WITHOUT changing  the size of the icon?  

@joshkimux My concern about enlarging the actual icon size is that -- although it's not an issue when the content is _inside_ the accordion -- it would be an issue when the icon is next to the h3 outside of the accordion. During internal team reviews, we had an issue where if it wrapped to the next line, the icon expanded the line height in a weird way. And Fareez did an excellent job in remedying that, and I don't want to go back to the same issue.

**Comment by joshkimux** [comment]
@jenniferlee-dsva Ahhh, got it! So this is in the case where it is next to the h3 outside of the accordion. Do we have an example of that anywhere, ideally on a mobile screen? I don't want to make a recommendation before giving a proper review of the design and code on different breakpoints and use cases.

**Comment by youngfreezyVA** [comment]
yup! you can visit this link, and then change to mobile in chrome dev tools:
https://staging.va.gov/coronavirus-veteran-frequently-asked-questions/#can-i-get-a-covid-19-vaccine-a-39606

**Comment by joshkimux** [comment]
<img width="1201" alt="Screen Shot 2021-05-13 at 6 12 36 PM" src="https://user-images.githubusercontent.com/14154792/118194089-cd34c580-b416-11eb-99d5-f806da0e4921.png">

@jenniferlee-dsva and @youngfreezyVA have we tried flexing the heading?

**Comment by jenniferlee-dsva** [comment]
Hi @joshkimux - I don't understand what "flexing the heading" means, so if that's some kind of coding term, I'll let Fareez answer that one.  

I think what you show above, from a UX pov, looks fine and would resolve the line wrap issue of line height weirdly blowing up. 

1. Would this be behavior only appear for  mobile or on desktop  as well? 
2. How would this affect the icon alignment with the text? (We've fine tuned so that the icon is precisely aligned to the top of the line height.) 
3. How would this affect alignment and UX of the activated state? (Where the "Copy linked" message displays; so on.)

If the latter two Qs are none or @youngfreezyVA  is able to make the same as we have now, your suggested solution sounds good to me. But I'd like Fareez to weigh in from FE pov. 

BTW, Josh, where did you see that question (in your screenshot)?  I don't see that particular text in staging at all. The deep link Fareez provided goes to a question called "Can I get a COVID-19 vaccine at VA?" not "How can I get...."

**Comment by joshkimux** [comment]
@jenniferlee-dsva I can answer some of these, but agree that ultimately @youngfreezyVA has the final word.

1. It could be for both (see video)

https://user-images.githubusercontent.com/14154792/118279959-b8e5dc80-b499-11eb-9c29-fae3ffe76802.mov

2. If we use a flexbox on the heading, the copy button will inherit the top margin from that heading and remain aligned with the text. That being said, increasing the actual size of the button itself can make it taller than the heading which can push text down. I have another possible way of addressing that below, but feel free to toss if it's too much of a lift.

https://user-images.githubusercontent.com/14154792/118280615-635dff80-b49a-11eb-8716-6ab26c27030c.mov

3. It wouldn't have any changes to the current state. The breaking you see in the suggested change exists in current state too.

**Suggested**

https://user-images.githubusercontent.com/14154792/118280814-8be5f980-b49a-11eb-91e8-22e57270ac9c.mov

**Current**

https://user-images.githubusercontent.com/14154792/118280964-b59f2080-b49a-11eb-9dc7-34ef546f5adc.mov


An alternative, perhaps more robust way of approaching this in a future iteration may be to split the "copy link" button apart into an anchor _link_ and an actual share link _button_. We can accommodate more folks by providing alternative ways of copying and sharing a link.

[I like this accessible example of an anchor link](https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/#what-are-anchor-links-exactly%3F) which is absolutely positioned off to the left and then flexed on smaller devices.

<img width="771" alt="Screen Shot 2021-05-14 at 9 24 58 AM" src="https://user-images.githubusercontent.com/14154792/118281719-8937d400-b49b-11eb-901f-b919d0847512.png">
Note: couldn't find the original share link icon, so please forgive the emoji usage

This would also resolve the following issues:
- We wouldn't have to worry about spacing or text alignment with the heading text anymore since we could use a native "#" with `sr-only` text or an `aria-label`.
- We can include text copy in the button so it's more clear to everyone and it can accommodate folks who use voice assistive technology to click buttons e.g. With a visible label, it will be clear what to read out for the AT to take action e.g. "Click share link button." In the current state, a link icon could be interpreted as, "Click link button" which may not activate the button if their best guess of what the icon is doesn't match the accessible name.
- We can beef up the size of the share link button without worrying about it causing alignment issues with the anchor link.
- Users wouldn't mistake the copy link button in the heading to behave similarly to anchor links they see elsewhere on the web
- Instead of providing pop up feedback about the link being copied, we could just update the already existing label within the copy share link button

**Comment by jenniferlee-dsva** [comment]
A lot to noodle on. Thanks @joshkimux . 

I think at one point we actually did consider placing the copy link icon in front of the text header. I don't remember why we ultimately decided against that - it may have had to do with the need to have it be used both inside and outside of accordions; also perhaps -- erroneously? -- we assumed it might not be 508 to have it come _before_ the text. 

To your concern about the links needing to indicate they are different from anchor links -- a**ctually they behave the same way.  The icon is simply a copying action.** No sharing action. And when a recipient gets the copied link and clicks on it, just like clicking on an anchor link, it takes them directly to that deep content. 

I like your simplified suggestion of moving the `copy link` icon in front of the header text (outside of accordions). Although we would still want to use the icon. We need to have consistency whether the link copying affordance is inside or outside accordions, since they are performing the same action and bc I think we'd be putting too much burden on users to process 2 different looking affordances that do the same thing.

**Comment by youngfreezyVA** [comment]
thanks @joshkimux ... I believe I tried making it a flex column, but it lead to the height of the overall h3 increasing (like in your screenshots). we may want to consolidate all this feedback in a separate ticket, as i'm not sure this is a defect per se since we are brainstorming the overall design

**Comment by jenniferlee-dsva** [comment]
My original question - is the tappable area increasable without enlarging the actual icon dimension? (So we can also avoid/not re-introduce line height and other issues.)  

This would be my preferred approach for addressing the touch target issue. 

Barring the above, if we have to enlarge the icon dimension, we'll  figure out a way to do this post-launch -- for both inside and outside accordions -- while not (re)introducing weird line height/wrapping issues. 

It seems like the flexbox videos look like it's possible (maybe the flex box is only around the heading text and not also including the icon?).

**Comment by joshkimux** [comment]
@jenniferlee-dsva 

> To your concern about the links needing to indicate they are different from anchor links -- actually they behave the same way. The icon is simply a copying action. 

Let me clarify a bit more! **The anchor _link_ simply anchors someone on a part of the page, it doesn't actually copy a link.** In other words, it just updates the browser's link to be more specific to the location on the page. For example, instead of `va.gov/covid`, you might click on an anchor link that's next to the heading "Covid Information" and get `va.gov/covid#covid-information`. 

This is a relatively common pattern on the web. I'd recommend trying it out on [Adam Silver's blog](https://adamsilver.io/blog/but-sometimes-buttons-look-like-links/) or seeing how to make it in [Amber Wilson's anchor link writeup](https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/#what-are-anchor-links-exactly%3F) (which got blessings from the legendary [Jeremy Keith](https://adactio.com/) himself).

**The action of "copying" requires javascript, which is why it becomes a _button_.** You're correct that there is no sharing action, that was just poor semantic choice from my part. The literal action would just be "Copy link".

That being said, they actually do different things despite sharing a similar user goal. It's giving more options to users to get to what they need essentially (copy a link) while packaging everything neatly in the design.

**Comment by joshkimux** [comment]
> My original question - is the tappable area increasable without enlarging the actual icon dimension? (So we can also avoid/not re-introduce line height and other issues.)

Not to my knowledge in the current implementation. In the suggested flexbox implementation, to @youngfreezyVA 's point, it will increase the total height of the `h3` (you can see that slightly below). It's noticeable, but not major and doesn't break the headings line height.

<img width="242" alt="Screen Shot 2021-05-14 at 11 35 43 AM" src="https://user-images.githubusercontent.com/14154792/118294250-85f71500-b4a8-11eb-9df1-36790752d63a.png">


> Barring the above, if we have to enlarge the icon dimension, we'll figure out a way to do this post-launch -- for both inside and outside accordions -- while not (re)introducing weird line height/wrapping issues.

Agreed, this is a post-launch defect ticket, and I think it deserves more time and thought to get right.

**Comment by jenniferlee-dsva** [comment]
Thank you for those clarifications @joshkimux . Much appreciated. 

@SKasimow and @youngfreezyVA - let's have a quick call on this after we've removed the launch blocker.

**Comment by SKasimow** [comment]
@youngfreezyVA .... Circling back on this...do we need a follow-up conversation? Or is there a clear approach to enlarging the icon (without creating other issues)?

CC @joshkimux @jenniferlee-dsva

**Comment by youngfreezyVA** [comment]
making the icon bigger is definitely going to cause a bunch of other
issues. I'll dig deeper when I have the chance and provide findings.

**Comment by joshkimux** [comment]
@SKasimow and @youngfreezyVA , agreed, making the icon bigger with the current design will likely cause unintended issues related to spacing.

I've documented a [possible fix on ticket 25277 for the full a11y audit.](https://github.com/department-of-veterans-affairs/va.gov-team/issues/25277)

**Comment by youngfreezyVA** [comment]
@joshkimux yes agreed, we probably need to go back to the drawing board w/ this one. wish this accessibility feedback was caught earlier in the collab cycle.

**Comment by joshkimux** [comment]
@youngfreezyVA absolutely, and I'm sorry I couldn't catch this one earlier in the process. 🙇 
Accessibility reviews follow the swiss cheese model where we can catch more issues the more we're involved 🧀 

To prevent something like this from repeating in the future, I'd encourage the team to collaborate with us a11y folks earlier and more often-- especially during wireframing. That should reduce the amount of remediation efforts down the road 🤝
